caddy.go

```diff
@@ -867,3 +867,19 @@ var (
 	// by default if no other file is specified.
 	DefaultConfigFile = "Caddyfile"
 )
+
+// contextKey is a value for use with context.WithValue. Inspired by the
+// net/http source.
+type contextKey struct {
+	name string
+}
```

Comment:
As is, the staticfiles package cannot use CtxKey because importing httpserver would result in an import cycle. My rationale for putting my contextKey type in the caddy package was not only to avoid the import cycle but also to make the predefined keys available to the caddytls package where HTTPChallengeHandler is implemented.

If you don't like the idea of putting CtxKey definition in the caddy package then how about a new package within caddyhttp? Then, httpserver and other middlewares (staticfiles, fastcgi, basicauth, etc.) could each import the package and have access to the predefined keys declared therein.

I have some local commits that prove it out, but wanted to get others' take before pushing.